### PR TITLE
Add Password Complexity Configuration

### PR DESCRIPTION
**Pull request by dmizelle**
A potential solution to #1963.

This commit does the following:

1. Moves the values for minimum and maximum password length into the
ClientAPI configuration struct.
2. Introduces a new struct representing the password complexity
requirements defined in dendrite-config.yml, with four options. Defaults
are compatible with what users probably expect out of synapse.
  * Minimum length, default of 8
  * Maximum length, default of 512
  * Minimum number of symbols, default of 0
  * Requiring mixed case toggle, default of false
3. Adds tests for the logic of validating passwords.

Signed-off-by: Devon Mizelle <dev@devon.so>

### Pull Request Checklist

* [x] I have added any new tests that need to pass to `sytest-whitelist` as specified in [docs/sytest.md](https://github.com/matrix-org/dendrite/blob/master/docs/sytest.md)
* [x] Pull request includes a [sign off](https://github.com/matrix-org/dendrite/blob/master/docs/CONTRIBUTING.md#sign-off)

### CODE REVIEW

File: clientapi/routing/register.go [hunk 1]
```diff
@@ -935,7 +958,7 @@ func RegisterAvailable(
 	}
 }
 
-func handleSharedSecretRegistration(userAPI userapi.UserInternalAPI, sr *SharedSecretRegistration, req *http.Request) util.JSONResponse {
+func handleSharedSecretRegistration(userAPI userapi.UserInternalAPI, sr *SharedSecretRegistration, passwordRequirements config.PasswordRequirements, req *http.Request) util.JSONResponse {
```
Comment (dmizelle) [inline clientapi/routing/register.go:961]: I'm a little concerned here. I'm not really sure if this shared secret registration should share the same requirements as user passwords. I'm happy to have it do the defaults instead (8 minlength & 512minlength and nothing else) regardless of what the user specifies in the config.

File: clientapi/routing/register_test.go [hunk 2]
```diff
@@ -208,3 +212,123 @@ func TestValidationOfApplicationServices(t *testing.T) {
...
+		MinNumberSymbols:  5,
+	}
+	var testCases = []struct {
+		name     string
+		config   config.PasswordRequirements
+		password string
+		expected *util.JSONResponse
+	}{
+		{
+			"default reject too short",
```
Comment (S7evinK) [inline clientapi/routing/register_test.go:234]: Minor nitpick: the tests are "hard" to read, without the fields specified.

File: clientapi/routing/register.go (hunk 1 above)
Comment (S7evinK) [reply clientapi/routing/register.go:961]: I think it's fine to use the same requirements, as shared secret registration isn't any different to a normal registration.

**Review by S7evinK — APPROVED** [review]
LGTM.

File: clientapi/routing/register_test.go (hunk 2 above)
Comment (neilalexander) [reply clientapi/routing/register_test.go:234]: Second this — can you please name the fields in the values?

File: clientapi/routing/register.go [hunk 3]
```diff
@@ -225,17 +227,38 @@ func validateApplicationServiceUsername(username string) *util.JSONResponse {
...
 	// https://github.com/matrix-org/synapse/blob/v0.20.0/synapse/rest/client/v2_alpha/register.py#L161
-	if len(password) > maxPasswordLength {
+	if len(password) > cfg.MaxPasswordLength {
 		return &util.JSONResponse{
 			Code: http.StatusBadRequest,
-			JSON: jsonerror.BadJSON(fmt.Sprintf("'password' >%d characters", maxPasswordLength)),
+			JSON: jsonerror.WeakPassword(fmt.Sprintf("'password' >%d characters", cfg.MaxPasswordLength)),
 		}
-	} else if len(password) > 0 && len(password) < minPasswordLength {
+	} else if len(password) > 0 && len(password) < cfg.MinPasswordLength {
```
Comment (kegsay) [inline clientapi/routing/register.go:237]: This fails when `len(password) == 0` as it'll fallthrough.

File: clientapi/routing/register.go [hunk 4]
```diff
@@ -225,17 +227,38 @@ func validateApplicationServiceUsername(username string) *util.JSONResponse {
...
+			JSON: jsonerror.WeakPassword(fmt.Sprintf("password too weak: min %d chars", cfg.MinPasswordLength)),
+		}
+	}
+
+	if cfg.MinNumberSymbols > 0 {
+		matches := passwordSymbols.FindAllStringIndex(password, -1)
+		if len(matches) < cfg.MinNumberSymbols {
+			return &util.JSONResponse{
+				Code: http.StatusBadRequest,
+				JSON: jsonerror.WeakPassword(fmt.Sprintf("password too weak: minimum %d symbols", cfg.MinNumberSymbols)),
```
Comment (kegsay) [inline clientapi/routing/register.go:249]: "Symbols" is a bit of a stretch when we are really talking about non-alphanumerics. E.g `ë` will be treated as a symbol which doesn't seem right.

File: clientapi/routing/register.go (hunk 1 above)
Comment (kegsay) [reply clientapi/routing/register.go:961]: Agreed, let's have a single consistent ruleset for passwords.

File: clientapi/routing/register_test.go [hunk 5]
```diff
@@ -208,3 +212,123 @@ func TestValidationOfApplicationServices(t *testing.T) {
...
+			&util.JSONResponse{
+				Code: http.StatusBadRequest,
+				JSON: jsonerror.WeakPassword(fmt.Sprintf("password too weak: min %d chars", defaults.MinPasswordLength)),
+			},
+		},
+		{
+			"default reject too long",
+			*defaults,
+			// len 600
+			"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
```
Comment (kegsay) [inline clientapi/routing/register_test.go:246]: `strings.Repeat("a", 600)` would be clearer.

File: clientapi/routing/register_test.go [hunk 6]
```diff
@@ -208,3 +212,123 @@ func TestValidationOfApplicationServices(t *testing.T) {
...
+				t.Errorf("expected password to be validated. got error: %s", response.JSON.(*jsonerror.MatrixError).Err)
+			} else if test.expected != nil && response == nil {
+				t.Errorf("expected password to fail, but was validated")
+			} else if test.expected != nil && test.expected.Code != response.Code {
+				t.Errorf("expected error code %d, got %d", test.expected.Code, response.Code)
+			} else if test.expected != nil && response != nil {
+				matrixError := response.JSON.(*jsonerror.MatrixError)
+				expectedError := test.expected.JSON.(*jsonerror.MatrixError)
+
+				if expectedError.Err != matrixError.Err {
```
Comment (kegsay) [inline clientapi/routing/register_test.go:324]: Please don't do this. Asserting error strings is really a pain, we have error codes for a reason, use that instead.

File: clientapi/routing/register_test.go [hunk 7]
```diff
@@ -208,3 +212,123 @@ func TestValidationOfApplicationServices(t *testing.T) {
...
+			"aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa",
+			&util.JSONResponse{
+				Code: http.StatusBadRequest,
+				JSON: jsonerror.WeakPassword(fmt.Sprintf("'password' >%d characters", defaults.MaxPasswordLength)),
+			},
+		},
+		{
+			"default allow long enough",
+			*defaults,
+			"thisisalongenoughpassword",
```
Comment (kegsay) [inline clientapi/routing/register_test.go:255]: This is brittle. If we change the default length of passwords then this test will fail because we're assuming the values. I see little benefit in testing this. By all means, check that:
```go
defaults := &config.PasswordRequirements{}
defaults.Defaults()
```
sets some sensible values (e.g min > 0 and max > min) but try to not bake in assumptions about the values.

File: clientapi/routing/register_test.go [hunk 8]
```diff
@@ -208,3 +212,123 @@ func TestValidationOfApplicationServices(t *testing.T) {
...
+				Code: http.StatusBadRequest,
+				JSON: jsonerror.WeakPassword("password must have uppercase and lowercase letters"),
+			},
+		},
+		{
+			"custom settings allow",
+			*custom,
+			"$0me_$up3r_$trong_P@ass",
+			nil,
+		},
```
Comment (kegsay) [inline clientapi/routing/register_test.go:306]: Add a test for an empty password please.

**Review by kegsay — CHANGES_REQUESTED** [review]

**Comment by dmizelle** [comment]
👋 thanks for the review y'all. I'll work through these later today.

File: clientapi/routing/register_test.go (hunk 2 above)
Comment (dmizelle) [reply clientapi/routing/register_test.go:234]: Does something like this work?

```
                        "default reject too short", // test name
                        *defaults,                  // password requirements
                        "foobar",                   // password value to test
                        &util.JSONResponse{ // expected response
                                Code: http.StatusBadRequest,
                                JSON: jsonerror.WeakPassword(fmt.Sprintf("password too weak: min %d chars", defaults.MinPasswordLength)),
                        },
```

File: clientapi/routing/register.go (hunk 3 above)
Comment (dmizelle) [reply clientapi/routing/register.go:237]: oh my, good catch. oops.

File: clientapi/routing/register_test.go (hunk 7 above)
Comment (dmizelle) [reply clientapi/routing/register_test.go:255]: removed this test. added a test in `setup/config/config_clientapi_test.go` to make sure the values are set correctly.

File: clientapi/routing/register_test.go (hunk 5 above)
Comment (dmizelle) [reply clientapi/routing/register_test.go:246]: what, I shouldn't scream in my tests? j/k :)

File: clientapi/routing/register_test.go (hunk 2 above)
Comment (dmizelle) [reply clientapi/routing/register_test.go:234]: oh! now i understand. okay, will do.

File: clientapi/routing/register.go (hunk 4 above)
Comment (dmizelle) [reply clientapi/routing/register.go:249]: There goes my anglophone-centric way of thinking here. :stuck_out_tongue:  I actually changed up this logic a bit to try and take into account non-latin characters / characters with symbols. Will mention you in the line where i switched it up.

File: clientapi/routing/register.go [hunk 9]
```diff
@@ -225,17 +224,49 @@ func validateApplicationServiceUsername(username string) *util.JSONResponse {
...
+
+	var (
+		symbolCount    = 0
+		uppercaseCount = 0
+		lowercaseCount = 0
+	)
+
+	for _, letter := range password {
+		switch {
+		case unicode.IsSymbol(letter) || unicode.IsPunct(letter):
```
Comment (dmizelle) [inline clientapi/routing/register.go:250]: @kegsay -- ended up changing this to use the `unicode` package instead, as I feel like its less anglocentric here. would love to get your thoughts.

**Comment by dmizelle** [comment]
Looks like I broke sytests on my branch. Will take a look later tonight.

File: clientapi/routing/register.go (hunk 3 above)
Comment (dmizelle) [reply clientapi/routing/register.go]: hey, so i cant actually deny empty passwords here. the most recent failure for sytest on buildkite comes from the AS tests:

https://buildkite.com/matrix-dot-org/dendrite/builds/5406

It looks like that in synapse itself, we "allow" empty passwords here:

https://github.com/matrix-org/synapse/blob/642a42eddece60afbbd5e5a6659fa9b939238b4a/synapse/handlers/password_policy.py#L52-L60

From what I understand (and I'm still learning, so excuse any dumb assumptions :) ) is that Application Services actually create users (specifically around IRC bridging) that have empty passwords.

I _think_ we should be denying account registrations with empty passwords, but that seems like it might get a little weird with application services.

I could edit the sytest tests to send a password along, but I'm sort of afraid that by merging this in with checking with length 0 passwords, that it could break application services in the future that might do this. :\

I hope I'm understanding correctly and that I haven't word salad'd :laughing:

**Comment by dmizelle** [comment]
(put a reply in one of the review comments above around AS and creating users)

File: clientapi/routing/register.go (hunk 3 above)
Comment (kegsay) [reply clientapi/routing/register.go]: We should allow empty passwords if it is an application service doing the registration, else we should reject.

Comment (dmizelle) [reply clientapi/routing/register.go]: @kegsay i keep coming back to trying to get this one to pass, but my gorilla brain can't really figure out how to differentiate between registration attempts that are application services and ones that are not.

is there a header or something that is set?

i want to figure out a way how to do this without breaking backwards compatibility (saying "hey, you now need to set XYZ thing to say you are an application service") but im sort of barking up weird trees.

Comment (S7evinK) [reply clientapi/routing/register.go]: [r.Type == authtypes.LoginTypeApplicationService](https://github.com/matrix-org/dendrite/blob/eadbfa7eea28beb5d706650ff9a158b68f37f521/clientapi/routing/register.go#L525) which could be passed to `validatePassword`. (Tested and now only `GET /register yields a set of flows` is still failing)

File: clientapi/routing/register.go (hunk 9 above)
Comment (kegsay) [reply clientapi/routing/register.go:250]: Seems reasonable.

File: clientapi/routing/register.go (hunk 3 above)
Comment (dmizelle) [reply clientapi/routing/register.go]: Oh hell yeah! Okay, I'm coming off a busy weekend, but I'll push a new commit soon.

Thanks a ton for the tip! <3

**Comment by danpe** [comment]
Amazing PR @dmizelle! Can we add another option to **restrict sequential and repetitive characters (e.g. 12345 or aaaaaa).** 
We can use the following regex for it: `/([a-zA-Z0-9])\1{2,}/`

**Comment by kegsay** [comment]
We recently updated our [contributing guidelines](https://github.com/matrix-org/dendrite/blob/main/docs/CONTRIBUTING.md#contribution-types). This PR is being closed because it isn't a feature we want to maintain going forwards. If there is a serious need for this, you can have a sidecar process to enforce password length and complexity.

**Comment by TruncatedDinoSour** [comment]
> We recently updated our [contributing guidelines](https://github.com/matrix-org/dendrite/blob/main/docs/CONTRIBUTING.md#contribution-types). This PR is being closed because it isn't a feature we want to maintain going forwards. If there is a serious need for this, you can have a sidecar process to enforce password length and complexity.

what even is a 'sidecar' process lmao
